observability/src/main/java/io/grpc/observability/interceptors/LogHelper.java

```diff
@@ -300,53 +317,85 @@ void logCancel(
         .setEventLogger(eventLogger)
         .setLogLevel(LogLevel.LOG_LEVEL_DEBUG)
         .setRpcId(rpcId);
-    sink.write(logEntryBuilder.build());
+    sink.write(logEntryBuilder.build(), locationTags, customTags);
   }
 
   GrpcLogRecord.Builder createTimestamp() {
     long nanos = timeProvider.currentTimeNanos();
     return GrpcLogRecord.newBuilder().setTimestamp(Timestamps.fromNanos(nanos));
   }
 
+  private static final Set<String> NEVER_INCLUDED_METADATA = new HashSet<>(
+      Collections.singletonList(
+          // grpc-status-details-bin is already logged in a field of the binlog proto
+          STATUS_DETAILS_KEY.name()));
+  private static final Set<String> ALWAYS_INCLUDED_METADATA = new HashSet<>(
+      Collections.singletonList(
+          "grpc-trace-bin"));
+
   static final class PayloadBuilder<T> {
     T payload;
     int size;
+    boolean truncated;
 
-    private PayloadBuilder(T payload, int size) {
+    private PayloadBuilder(T payload, int size, boolean truncated) {
       this.payload = payload;
       this.size = size;
+      this.truncated = truncated;
     }
   }
 
   // TODO(dnvindhya): Create a unit test for the metadata conversion
-  static PayloadBuilder<GrpcLogRecord.Metadata.Builder> createMetadataProto(Metadata metadata) {
+  static PayloadBuilder<GrpcLogRecord.Metadata.Builder> createMetadataProto(Metadata metadata,
+      int maxHeaderBytes) {
     checkNotNull(metadata, "metadata");
+    checkArgument(maxHeaderBytes >= 0,
+        "maxHeaderBytes must be non negative");
     GrpcLogRecord.Metadata.Builder metadataBuilder = GrpcLogRecord.Metadata.newBuilder();
     // This code is tightly coupled with io.grpc.observabilitylog.v1.GrpcLogRecord.Metadata's
     // implementation
     byte[][] serialized = InternalMetadata.serialize(metadata);
+    boolean truncated = false;
     int totalMetadataBytes = 0;
     if (serialized != null) {
-      int singleMetadataEntryBytes = 0;
       // Calculate bytes for each GrpcLogRecord.Metadata.MetadataEntry
       for (int i = 0; i < serialized.length; i += 2) {
         String key = new String(serialized[i], Charsets.UTF_8);
         byte[] value = serialized[i + 1];
-        singleMetadataEntryBytes = totalMetadataBytes + key.length() + value.length;
+        if (NEVER_INCLUDED_METADATA.contains(key)) {
```

Comment:
Haven't figured out the significance of this `NEVER_INCLUDED_METADATA` and where is the requirement coming from. The comment there talks about binlog